### ISSUE TITLE
PMTU reduction causes packet loss

### ISSUE THREAD

**Issue by DevinCarr** [quic-go/quic-go#3955]
The quic-go library uses an MTU discovery process after the handshake is completed between the client and the server (https://github.com/quic-go/quic-go/blob/master/mtu_discoverer.go). This code does not handle the case that the PMTU is lowered below the "current" byte count.

If you have a long-lived QUIC connection between a server and a client and the PMTU is adjusted to a lower value than previous, QUIC packets that were sending with full payloads will begin to get dropped.
We noticed this in cloudflared that if the connection is sending ping frames but no other payloads, a PMTU reduction will not cause any issues as the ping frames are always smaller than the PMTU (usually under 100 bytes). But, when the client-side attempts to send payloads that would normally fill the QUIC packet size, we began to see an increase in the packet loss metric. We investigated further into the MTU discovery code in quic-go and noticed that once the PMTU is established, it will never be lowered: https://github.com/quic-go/quic-go/blob/master/mtu_discoverer.go#L109

Here is a repro of the issue: https://github.com/DevinCarr/pmtu-failure-quic-go-repro
_note_: it does require two machines to repro as you will need to adjust the MTU on the client side sending the payload.

Would there be some way that quic-go would be able to adjust this PMTU after the connection has been established and the PMTU has been calculated?

As a remediation, I thought that once the PMTU is established, quic-go could reduce the interval at which it evaluates the PMTU by attempting to send a ping packet of the current size to validate at least that the current PMTU has not been adjusted along the route.

Attempting something like this led me to the code which controls the congestion control: https://github.com/quic-go/quic-go/blob/master/internal/congestion/cubic_sender.go#L308
Reducing the PMTU via the `mtuDiscoverer` causes this panic. Is there a reason why the congestion control cannot reduce the datagram size? (I apologize in advance for my ignorance on the QUIC congestion control and how it's implemented here)

Finally, beyond leveraging [`DisablePathMTUDiscovery`](https://github.com/quic-go/quic-go/blob/master/interface.go#L320), is there any mechanism that we would be able to account for this issue?
Sadly, this configuration is currently our only lever beyond restarting the connection and establishing the PMTU again for the new connection.

**Comment by marten-seemann** [comment]
Hi @DevinCarr, thank you for your detailed issue.

The short answer is: The current implementation isn't supposed to deal with reduction of the PMTU. The assumption is that the PMTU never decreases, and we took some shortcuts when implementing RFC 8899. That's why we have the panic in the congestion controller.

It sounds like you're running into a case where this assumption doesn't hold any more. Do you know why the PMTU changes?

> As a remediation, I thought that once the PMTU is established, quic-go could reduce the interval at which it evaluates the PMTU by attempting to send a ping packet of the current size to validate at least that the current PMTU has not been adjusted along the route.

I think we can do better than this. We don't need to send any probe packets with the current size, we're already sending plenty of packets (assuming we have application data to send).
The situation we need to react to is when a longer train of (full-size) packets is lost. I'm wondering if we should hook into the PTO logic of RFC 9002: instead of sending 2 full-size probe packets, we could send one full-size packet and one packet that's effectively a probe packet (although it could contain application data) with a smaller size (1200 bytes?). In case the small packet makes it through, we could then restart PMTUD. We could also decide to not declare persistent congestion in that case.

What do you think? Does this make sense? Would you be interested in submitting a patch? Please let me know, I can give you some pointers. For example, it might make sense to move the `mtuDiscoverer` into the `ackhandler` package.

**Comment by joliveirinha** [comment]
Thanks @marten-seemann for the quick response.

> It sounds like you're running into a case where this assumption doesn't hold any more. Do you know why the PMTU changes?

we are still investigating this, but when you start working with anycast networks, plus L4 load balancers that may wrap IP packets with an extra encapsulation in some cases, it may result in the reduction of end-2-end PMTU, even if for a small amount of bytes.

This is clearly something that is very rate, but that it can happen.

> What do you think? Does this make sense? Would you be interested in submitting a patch? Please let me know, I can give you some pointers. For example, it might make sense to move the mtuDiscoverer into the ackhandler package.

We would need to review the RFC and investigate the code in more detail. I think this is something that we want to address, so probably we will be doing a patch for this in the future. 
We would like to setup a meeting with you if possible to discuss the details if necessary once we have a proposal in our side.
WDYT?

Also, somewhat related to this topic but unrelated to this issue. Can you take a look at https://github.com/quic-go/quic-go/issues/3300 ? We would like to merge a change so we can stop having our own fork.

Thanks!

**Comment by marten-seemann** [comment]
> We would like to setup a meeting with you if possible to discuss the details if necessary once we have a proposal in our side.
> WDYT?

Sure! Will you be at the IETF meeting next week by any chance? We could meet up there. Otherwise, happy to schedule a call for the week after. Can you send me an email so we can discuss the details?

**Comment by DevinCarr** [comment]
Sadly none of us will be attending the IETF meeting next week, sorry we can't meetup in person!

I started work on this instead from the approach of the `max_udp_payload_size` since that is a bounding factor on-top of the `MaxDatagramSize` and would work better for our immediate concerns with the PMTU issue.

My biggest concern with the change so far is definitely the fact that `MaxPacketBufferSize` is used and assumed a constant. With adjusting the UDP payload size, this would not be the case, at least for the case that a user sets the `MaxUDPPayloadSize` greater than the `MaxPacketBufferSize`. When this occurs the full payload would not be read into the buffer and would fail payload decryption.

So there needs to be an adjustment where the buffers that are used for a connection are aware of the max (short|long)header+payload size and can at least increase to meet up to certain values. Likely this will only need to be considered for much larger packet sizes, and as I mentioned in the PR, could probably be just statically increased to a large value of perhaps 1552 (1600 - UDP header - IP header). This would mean that users would be limited to increase their payload size to a maximum value of 1552 and anything beyond would be clipped to that value. This is all assuming that you don't want to provide scaling buffers per connection that can go up to 65527 (or higher).

Let me know what you think and maybe we can iterate on this further!

**Comment by marten-seemann** [comment]
There’s an additional complication here. Unbeknownst to either endpoint, the networking might be employing some sort of ECMP, where some of the paths support a higher MTU than other paths.

Running PMTUD will lead to an increased packet loss rate, but might not lead to a PTO, since some packets will always make it through. See https://datatracker.ietf.org/doc/html/rfc4821#section-7.8 for some text about that.

We should probably have some detection based on high packet loss rates, but unfortunately that means coming up with some useful heuristics. It’s easy enough for an endpoint that’s continuously sending (>10% over the last N packets), but it gets more tricky once the endpoint is only intermittently sending packets.

**Comment by joliveirinha** [comment]
> We should probably have some detection based on high packet loss rates, but unfortunately that means coming up with some useful heuristics. It’s easy enough for an endpoint that’s continuously sending (>10% over the last N packets), but it gets more tricky once the endpoint is only intermittently sending packets.

True. In our case, what we are seeing is that for some reason, which are still trying to understand, after some time, the PTMU changes and all packets sent with high packet size (close to the PTMU initially calculated), they will get dropped.
So, having this logic of > 10% of last N packets sent, could help us detect this. But I think for it to work correctly, we would need to track that percentage per packet size bucket, thus increasing the probability of us triggering the heuristic of the 10%.
It's not perfect, that's true...

Another alternative, would be for the mtu discovery to also listen for ICMP PTB messages. I guess this approach for PTMU was not initially developed because ICMP blackhole is a thing, and it was safer to implement the current approach, which doesn't rely on it. 
But what if we had something hybrid? We still use the current approach, but use ICMP PTB to decrease the MTU. 

If users are in a environment that is subject to PTMU changes and also have ICMP blackhole, then this wouldn't solve it, but I guess at that point we are entering on the realm of how far we should go on this support.

The only limitation I see is that I am not sure how the ICMP package works for listening PTB packets work for all the env (eg. windows, ...). But this is something that could be investigated.

**Comment by marten-seemann** [comment]
> True. In our case, what we are seeing is that for some reason, which are still trying to understand, after some time, the PTMU changes and all packets sent with high packet size (close to the PTMU initially calculated), they will get dropped.
> So, having this logic of > 10% of last N packets sent, could help us detect this. But I think for it to work correctly, we would need to track that percentage per packet size bucket, thus increasing the probability of us triggering the heuristic of the 10%.

It might make sense to only count packets above and below 1200 bytes. QUIC requires 1200 bytes, so if packets smaller than that don't make it through, the path is incapable of carrying QUIC in the first place.

> Another alternative, would be for the mtu discovery to also listen for ICMP PTB messages. I guess this approach for PTMU was not initially developed because ICMP blackhole is a thing, and it was safer to implement the current approach, which doesn't rely on it.

That's an optimization, but you can never rely on ICMP making it through, unfortunately.
